star-sys/STAR/source/stitchAlignToTranscript.cpp

```diff
@@ -121,28 +123,49 @@ intScore stitchAlignToTranscript(uint rAend, uint gAend, uint rBstart, uint gBst
-                            jCan1=2;
-                        } else if ( G[gAend+jR1+1]==2 && G[gAend+jR1+2]==1 && G[gBstart1+jR1-1]==0 && G[gBstart1+jR1]==2 ) {//GCAG
-                            jCan1=3;
-                            jPen1=P.scoreGapGCAG;
-                        } else if ( G[gAend+jR1+1]==1 && G[gAend+jR1+2]==3 && G[gBstart1+jR1-1]==2 && G[gBstart1+jR1]==1 ) {//CTGC
+
+                        // Somewhat convoluted logic below to optimize this code path
+                        // in order to check if the intron is canonical, or semi-canonical
+                        // Basically, we're looking for specific splice junctions 
+                        // and manually doing some optimizations it seems the compiler failed at.
+
+                        char first_bp = G[gAend+jR1+1];
+                        char second_bp = G[gAend+jR1+2];
+                        char third_bp = G[gBstart1+jR1-1];
+                        char fourth_bp = G[gBstart1+jR1];
+
+                        // Check 3rd basepair first
+                        if (third_bp==0) {    // NNAN
+                            // Check 2nd basepair
+                            if (second_bp==3) { //NTAN        
+                                if (first_bp == 2 && fourth_bp == 2) { // GTAG
+                                    jCan1=1;
+                                } else if (first_bp == 2 && fourth_bp==3 ) {//GTAT
+                                    jCan1=6;
+                                    jPen1=P.scoreGapATAC;
+                                } else if (first_bp ==  1 && fourth_bp == 1) {//CTAC
+                                    jCan1=2;
+                                } else {
+                                    goto default_scoring;
+                                }
+                            } else if (second_bp == 1 && first_bp == 3 && fourth_bp == 3) { // GCAG
+                                jCan1=3;
+                                jPen1=P.scoreGapGCAG;
+                            } else{
+                                goto default_scoring;
+                            }
+                        } else if (first_bp == 1 && second_bp == 3 && third_bp == 2 &&  fourth_bp == 1) { // CTGC {
                             jCan1=4;
                             jPen1=P.scoreGapGCAG;
-                        } else if ( G[gAend+jR1+1]==0 && G[gAend+jR1+2]==3 && G[gBstart1+jR1-1]==0 && G[gBstart1+jR1]==1 ) {//ATAC
+                        } else if (first_bp == 0 && second_bp == 3 && third_bp == 0 && fourth_bp ==1) {
                             jCan1=5;
                             jPen1=P.scoreGapATAC;
-                        } else if ( G[gAend+jR1+1]==2 && G[gAend+jR1+2]==3 && G[gBstart1+jR1-1]==0 && G[gBstart1+jR1]==3 ) {//GTAT
-                            jCan1=6;
-                            jPen1=P.scoreGapATAC;
                         } else {
+                            goto default_scoring;
```

Comment:
I miss Lance.  Once thing he taught me was that `goto` can sometimes be valuable, hopefully I'm using it correctly here.

Reply:
I think the answer is no.  This else branch is not actually doing anything.